README.md
<p align="center">
<img width="600px" src="docs/src/assets/logo.png"/>
</p>

# DBnomics.jl

## DBnomics Julia client

This package provides you access to DBnomics data series. DBnomics is an open-source project with the goal of aggregating the world's economic data in one location, free of charge to the public. DBnomics covers hundreds of millions of series from international and national institutions (Eurostat, World Bank, IMF, ...).

To use this package, you have to provide the codes of the provider, dataset and series you want. You can retrieve them directly on the <a href="https://db.nomics.world/" target="_blank">website</a>.

To install `DBnomics.jl`, go to the package manager with `]` :

```julia
add DBnomics
```
or install the github version with :

```julia
add https://github.com/s915/DBnomics.jl
```

All the functions, and their names, are derived from the R package <a href="https://github.com/dbnomics/rdbnomics" target="_blank"><b>rdbnomics</b></a> which I also maintain.

## Examples
Fetch time series by `ids` :
```julia
# Fetch one series from dataset 'Unemployment rate' (ZUTN) of AMECO provider :
df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN");

# Fetch two series from dataset 'Unemployment rate' (ZUTN) of AMECO provider :
df2 = rdb(ids = ["AMECO/ZUTN/EA19.1.0.0.0.ZUTN", "AMECO/ZUTN/DNK.1.0.0.0.ZUTN"]);

# Fetch two series from different datasets of different providers :
df3 = rdb(ids = ["AMECO/ZUTN/EA19.1.0.0.0.ZUTN", "IMF/BOP/A.FR.BCA_BP6_EUR"]);
```

In the event that you only use the argument `ids`, you can drop it and run :
```julia
df1 = rdb("AMECO/ZUTN/EA19.1.0.0.0.ZUTN");
```

Fetch time series by `mask` :
```julia
# Fetch one series from dataset 'Balance of Payments' (BOP) of IMF :
df1 = rdb("IMF", "BOP", mask = "A.FR.BCA_BP6_EUR");

# Fetch two series from dataset 'Balance of Payments' (BOP) of IMF :
df2 = rdb("IMF", "BOP", mask = "A.FR+ES.BCA_BP6_EUR");

# Fetch all series along one dimension from dataset 'Balance of Payments' (BOP) of IMF :
df3 = rdb("IMF", "BOP", mask = "A..BCA_BP6_EUR");

# Fetch series along multiple dimensions from dataset 'Balance of Payments' (BOP) of IMF :
df4 = rdb("IMF", "BOP", mask = "A.FR.BCA_BP6_EUR+IA_BP6_EUR");
```

In the event that you only use the arguments `provider_code`, `dataset_code` and `mask`, you can drop the name `mask` and run :
```julia
df1 = rdb("IMF", "BOP", "A.FR.BCA_BP6_EUR");
```

Fetch time series by `dimensions` :
```julia
# Fetch one value of one dimension from dataset 'Unemployment rate' (ZUTN) of AMECO provider :
df1 = rdb("AMECO", "ZUTN", dimensions = Dict(:geo => "ea12"));
# or
df1 = rdb("AMECO", "ZUTN", dimensions = (geo = "ea12",));

# Fetch two values of one dimension from dataset 'Unemployment rate' (ZUTN) of AMECO provider :
df2 = rdb("AMECO", "ZUTN", dimensions = Dict(:geo => ["ea12", "dnk"]));
# or
df2 = rdb("AMECO", "ZUTN", dimensions = (geo = ["ea12", "dnk"],));

# Fetch several values of several dimensions from dataset 'Doing business' (DB) of World Bank :
df3 = rdb("WB", "DB", dimensions = Dict(:country => ["DZ", "PE"], :indicator => ["ENF.CONT.COEN.COST.ZS", "IC.REG.COST.PC.FE.ZS"]));
# or
df3 = rdb("WB", "DB", dimensions = (country = ["DZ", "PE"], indicator = ["ENF.CONT.COEN.COST.ZS", "IC.REG.COST.PC.FE.ZS"]));
```

Fetch time series with a `query`:
```julia
# Fetch one series from dataset 'WEO:2019-10 by countries' (WEO:2019-10) of IMF provider:
df1 = rdb("IMF", "WEO:2019-10", query = "France current account balance percent");

# Fetch series from dataset 'WEO:2019-10 by countries' (WEO:2019-10) of IMF provider:
df2 = rdb("IMF", "WEO:2019-10", query = "current account balance percent");
```

Fetch one series from the dataset 'Doing Business' of WB provider with the link:
```julia
df1 = rdb(api_link = "https://api.db.nomics.world/v22/series/WB/DB?dimensions=%7B%22country%22%3A%5B%22FR%22%2C%22IT%22%2C%22ES%22%5D%7D&q=IC.REG.PROC.FE.NO&observations=1&format=json&align_periods=1&offset=0&facets=0");
```

In the event that you only use the argument `api_link`, you can drop the name and run:
```julia
df1 = rdb("https://api.db.nomics.world/v22/series/WB/DB?dimensions=%7B%22country%22%3A%5B%22FR%22%2C%22IT%22%2C%22ES%22%5D%7D&q=IC.REG.PROC.FE.NO&observations=1&format=json&align_periods=1&offset=0&facets=0");
```

Fetch the available datasets of a provider
```julia
# Example with the IMF datasets:
df_datasets = rdb_datasets("IMF");

# Example with the IMF and BDF datasets:
df_datasets = rdb_datasets(["IMF", "BDF"]);
```

In the event that you only request the datasets for one provider, if you define
`simplify = true`, then the result will be a `DataFrame` not a `Dict`.
```julia
df_datasets = rdb_datasets("IMF", simplify = true);
```

Fetch the possible dimensions of available datasets of a provider
```julia
# Example for the dataset WEO:2019-10 of the IMF:
df_dimensions = rdb_dimensions("IMF", "WEO:2019-10");
```

In the event that you only request the dimensions for one dataset for one
provider, if you define `simplify = true`, then the result will be a `Dict` of
`DataFrame`s not a nested `Dict`.
```julia
df_dimensions = rdb_dimensions("IMF", "WEO:2019-10", simplify = true);
```

Fetch the number of series of available datasets of a provider
```julia
# Example for the dataset WEOAGG:2019-10 of the IMF:
df_series = rdb_series("IMF", "WEOAGG:2019-10");

# With dimensions
df_series = rdb_series("IMF", "WEO:2019-10", dimensions = Dict(Symbol("weo-country") => "AGO"));
df_series = rdb_series("IMF", "WEO:2019-10", dimensions = Dict(Symbol("weo-subject") => "NGDP_RPCH"), simplify = true);

# With a query
df_series = rdb_series("IMF", "WEO:2019-10", query = "ARE");
df_series = rdb_series("IMF", ["WEO:2019-10", "WEOAGG:2019-10"], query = "NGDP_RPCH");
```

:warning: We ask the user to use this function parsimoniously because there are a huge amount
of series per dataset. Please only fetch for one dataset if you need it or
visit the website [https://db.nomics.world](https://db.nomics.world).  

## Proxy configuration
When using the functions `rdb` or `rdb_...`, if you come across an error concerning your internet connection, you can get round this situation by :

1. configuring **curl** of the function `HTTP.get` or `HTTP.post` to use a specific and authorized proxy.

2. using the functions `readlines` and `download` if you have problem with `HTTP.get`.

### Configure **curl** to use a specific and authorized proxy
In **DBnomics.jl**, by default the function `HTTP.get` or `HTTP.post` are used to fetch the data. If a specific proxy must be used, it is possible to define it permanently with the package global variable `curl_config` or on the fly through the argument `curl_config`. In that way the object is passed to the keyword arguments of the function `HTTP.get` or `HTTP.post`.  
To see the available parameters, visit the website <a href="https://curl.haxx.se/libcurl/c/curl_easy_setopt.html" target="_blank">https://curl.haxx.se/libcurl/c/curl_easy_setopt.html</a>.  
Once they are chosen, you define the curl object as follows :
```julia
h = Dict(:proxy => "http://<proxy>:<port>");
```

Regarding the functioning of **HTTP.jl**, you might need to modify another option to change the *db/editor.nomics.world* url from *https://* to *http://* (see https://github.com/JuliaWeb/HTTP.jl/pull/390) :
```julia
DBnomics.options("secure", false);
```

#### Set the connection up for a session
The curl connection can be set up for a session by modifying the following package option :
```julia
DBnomics.options("curl_config", h);
```
After configuration, just use the standard functions of **DBnomics.jl** e.g. :
```julia
df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN");
```
This option of the package can be disabled with :
```julia
DBnomics.options("curl_config", nothing);
```

#### Use the connection only for a function call
If a complete configuration is not needed but just an "on the fly" execution, then use the argument `curl_config` of the functions `rdb` and `rdb_...` :
```julia
df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN", curl_config = h);
```

### Use the standard functions `readlines` and `download`
To retrieve the data **DBnomics.jl** can also use the standard functions `readlines` and `download`.

#### Set the connection up for a session
To activate this feature for a session, you need to enable an option of the package :
```julia
DBnomics.options("use_readlines", true);
```
And then use the standard function as follows :
```julia
df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN");
```
This configuration can be disabled with :
```julia
DBnomics.options("use_readlines", false);
```

#### Use the connection only for a function call
If you just want to do it once, you may use the argument `use_readlines` of the functions `rdb` and `rdb_...` :
```julia
df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN", use_readlines = true);
```

## Transform time series with filters
The **DBnomics.jl** package can interact with the *Time Series Editor* of DBnomics to transform time series by applying filters to them.  
Available filters are listed on the filters page [https://editor.nomics.world/filters](https://editor.nomics.world/filters).

Here is an example of how to proceed to interpolate two annual time series with a monthly frequency, using a spline interpolation:

```julia
filters = Dict(:code => "interpolate", :parameters => Dict(:frequency => "monthly", :method => "spline"));

df = rdb(ids = ["AMECO/ZUTN/EA19.1.0.0.0.ZUTN", "AMECO/ZUTN/DNK.1.0.0.0.ZUTN"], filters = filters);
```

If you want to apply more than one filter, the `filters` argument will be a Tuple of valid filters:
```julia
filter1 = Dict(:code => "interpolate", :parameters => Dict(:frequency => "monthly", :method => "spline"));
filter2 = Dict(:code => "aggregate", :parameters => Dict(:frequency => "bi-annual", :method => "end_of_period"));
filters = (filter1, filter2);

df = rdb(ids = ["AMECO/ZUTN/EA19.1.0.0.0.ZUTN", "AMECO/ZUTN/DNK.1.0.0.0.ZUTN"], filters = filters);
```

The `DataFrame` columns change a little bit when filters are used. There are two new columns:

- `period_middle_day`: the middle day of `original_period` (can be useful when you compare graphically interpolated series and original ones).
- `filtered` (boolean): `true` if the series is filtered, `false` otherwise.

The content of two columns are modified:

- `series_code`: same as before for original series, but the suffix `_filtered` is added for filtered series.
- `series_name`: same as before for original series, but the suffix ` (filtered)` is added for filtered series.

## Transform the `DataFrame` object into a `TimeArray` object
For some analysis, it is more convenient to have a `TimeArray` object instead of a `DataFrame` object. To transform
it, you can use the following functions :
```julia
using DBnomics
using DataFrames
using TimeSeries

function to_namedtuples(x::DataFrames.DataFrame)
    nm = names(x)
    vl = try
        [x[!, col] for col in names(x)]
    catch
        [x[:, col] for col in names(x)]
    end
    nm = tuple(Symbol.(nm)...)
    vl = tuple(vl...)

    NamedTuple{nm}(vl)
end

function to_timeseries(
    x::DataFrames.DataFrame,
    index = :period, variable = :series_code, value = :value
)
    x = unstack(x, index, variable, value)
    x = to_namedtuples(x)
    x = TimeArray(x, timestamp = index)
    x
end

rdb("IMF", "BOP", mask = "A.FR+ES.BCA_BP6_EUR")
#> 162×18 DataFrame. Omitted printing of 12 columns
#> │ Row │ @frequency │ FREQ   │ Frequency │ INDICATOR   │ Indicator                          │ REF_AREA │
#> │     │ String     │ String │ String    │ String      │ String                             │ String   │
#> ├─────┼────────────┼────────┼───────────┼─────────────┼────────────────────────────────────┼──────────┤
#> │ 1   │ annual     │ A      │ Annual    │ BCA_BP6_EUR │ Current Account, Total, Net, Euros │ ES       │
#> │ 2   │ annual     │ A      │ Annual    │ BCA_BP6_EUR │ Current Account, Total, Net, Euros │ ES       │
#> │ ... │ ...        │ ...    │ ...       │ ...         │ ...                                │ ...      │
#> │ 161 │ annual     │ A      │ Annual    │ BCA_BP6_EUR │ Current Account, Total, Net, Euros │ FR       │
#> │ 162 │ annual     │ A      │ Annual    │ BCA_BP6_EUR │ Current Account, Total, Net, Euros │ FR       │

to_timeseries(rdb("IMF", "BOP", mask = "A.FR+ES.BCA_BP6_EUR"))
#> 81×2 TimeArray{Union{Missing, Float64},2,Date,Array{Union{Missing, Float64},2}} 1940-01-01 to 2020-01-01
#> │            │ A.ES.BCA_BP6_EUR │ A.FR.BCA_BP6_EUR │
#> ├────────────┼──────────────────┼──────────────────┤
#> │ 1940-01-01 │ missing          │ missing          │
#> │ 1941-01-01 │ missing          │ missing          │
#> │ 1942-01-01 │ missing          │ missing          │
#> │ ...        │ ...              │ ...              │
#> │ 2019-01-01 │ 24899.0          │ -16239.4         │
#> │ 2020-01-01 │ missing          │ missing          │
```

## P.S.
Visit <a href="https://db.nomics.world/" target="_blank">https://db.nomics.world/</a> !


NEWS.md
# Version 0.3.2

* Modification of the regular expression for the float values in the function `value_to_float!()`. From `Ref(r"^[0-9]*[[:blank:]]*[0-9]+\.*[0-9]*$")` to `Ref(r"^[0-9]*[[:blank:]]*[0-9]+\.*[0-9]*e*[0-9]*$")` (@tlorans, issue "Error with value_to_float! function #11").
* Add a compat entry for the DataFrames package : 1.2.

# Version 0.3.1

* Change the compat entry for the DataFrames package to 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.20, 0.21, 0.22, 1.0, 1.1.

# Version 0.3.0

* New argument 'query' and 'api_link' for function `rdb()`.
* `rdb_by_api_link()` is deprecated.
* New function `rdb_datasets()` to request the available datasets of the providers.
* New function `rdb_dimensions()` to request the list of the dimensions of the available datasets of the providers.
* New function `rdb_series()` to request the list of the series of the available datasets of the providers.
* New internal function `dot_rdb()`.

# Version 0.2.0

* New filters tool from <https://editor.nomics.world/filters>.
* If the retrieved dataset contains columns with codes (like ISO codes,
  geographic codes, ...), then correspondences are performed to translate
  these codes if possible.

# Version 0.1.0

* First release.


src/rdb.jl
"""
    rdb(
        provider_code::Union{Array, String, Nothing} = nothing,
        dataset_code::Union{String, Nothing} = nothing,
        mask_arg::Union{String, Nothing} = nothing;
        ids::Union{Array, String, Nothing} = nothing,
        dimensions::Union{Dict, NamedTuple, String, Nothing} = nothing,
        mask::Union{String, Nothing} = nothing,
        query::Union{String, Nothing} = nothing,
        api_link::Union{String, Nothing} = nothing,
        use_readlines::Bool = DBnomics.use_readlines,
        curl_config::Union{Nothing, Dict, NamedTuple} = DBnomics.curl_config,
        filters::Union{Nothing, Dict, Tuple} = DBnomics.filters,
        kwargs...
    )

`rdb` downloads data series from [DBnomics](https://db.nomics.world/) using shortcuts
like `ids`, `dimensions`, `mask`, `query` or using an `api_link`.

This function gives you access to hundreds of millions data series from
[DBnomics](https://db.nomics.world/). The code of each series is given on the
[DBnomics](https://db.nomics.world/) website.

In the event that the shortcut `ids` is used then the argument name can be
dropped and the `ids` will be passed through `provider_code`.

If only the argument `api_link` is provided, then the argument name can be dropped.
The string is directly passed to `api_link`.

In the same way, if only `provider_code`, `dataset_code` and `mask` are used then
the arguments names can be dropped. The `mask` will be passed through `mask_arg`.

By default, the function returns a `DataFrame`.

# Arguments
- `provider_code::Union{Array, String, Nothing} = nothing`: DBnomics code of the
  provider.
- `dataset_code::Union{String, Nothing} = nothing`: DBnomics code of the dataset.
- `mask_arg::Union{String, Nothing} = nothing`: DBnomics code of one or several masks in
  the specified provider and dataset. It is used if the arguments names are not given.
- `ids::Union{Array, String, Nothing} = nothing`: DBnomics code of one or several series.
- `dimensions::Union{Dict, NamedTuple, String, Nothing} = nothing`: DBnomics code of one
  or several dimensions in the specified provider and dataset. If it is a `Dict` or a
  `NamedTuple`, then then function `json` (from the package **JSON.jl**) is applied to
  generate the json object.
- `mask::Union{String, Nothing} = nothing`: DBnomics code of one or several masks in
  the specified provider and dataset.
- `query::Union{String, Nothing} = nothing`: A query to filter/select series from a
  provider's dataset.
- `api_link::Union{String, Nothing} = nothing`: DBnomics API link of the search.
  It should starts with `http://` or `https://`.
- `use_readlines::Bool = DBnomics.use_readlines`: (default `false`) If `true`, then
  the data are requested and read with the function `readlines`.
- `curl_config::Union{Nothing, Dict, NamedTuple} = DBnomics.curl_config`: (default `nothing`)
  If not `nothing`, it is used to configure a proxy connection. This
  configuration is passed to the keyword arguments of the function `HTTP.get` or `HTTP.post` of
  the package **HTTP.jl**.
- `filters::Union{Nothing, Dict, Tuple} = DBnomics.filters`: (default `nothing`)
  This argument must be a `Dict` for one filter because the function `json` of the
  package **JSON.jl** is used before sending the request to the server. For multiple
  filters, you have to provide a `Tuple` of valid filters (see examples).
  A valid filter is a `Dict` with a key `code` which value is a character string,
  and a key `parameters` which value is a `Dict` with keys `frequency`
  and `method` or `nothing`.
- `kwargs...`: Keyword arguments to be passed to `HTTP.get` or `HTTP.post`.

# Examples
```jldoctest
## By ids
julia> df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN");
# or
julia> df1 = rdb("AMECO/ZUTN/EA19.1.0.0.0.ZUTN");

julia> df2 = rdb(ids = ["AMECO/ZUTN/EA19.1.0.0.0.ZUTN", "AMECO/ZUTN/DNK.1.0.0.0.ZUTN"]);

julia> df3 = rdb(ids = ["AMECO/ZUTN/EA19.1.0.0.0.ZUTN", "IMF/BOP/A.FR.BCA_BP6_EUR"]);


## By dimensions
julia> df1 = rdb("AMECO", "ZUTN", dimensions = Dict(:geo => "ea12"));
# or
julia> df1 = rdb("AMECO", "ZUTN", dimensions = (geo = "ea12",));

julia> df2 = rdb("AMECO", "ZUTN", dimensions = Dict(:geo => ["ea12", "dnk"]));
# or
julia> df2 = rdb("AMECO", "ZUTN", dimensions = (geo = ["ea12", "dnk"],));

julia> dim = Dict(:country => ["DZ", "PE"], :indicator => ["ENF.CONT.COEN.COST.ZS", "IC.REG.COST.PC.FE.ZS"]);
julia> df3 = rdb("WB", "DB", dimensions = dim);
# or
julia> dim = (country = ["DZ", "PE"], indicator = ["ENF.CONT.COEN.COST.ZS", "IC.REG.COST.PC.FE.ZS"]);
julia> df3 = rdb("WB", "DB", dimensions = dim);


## By mask
julia> df1 = rdb("IMF", "BOP", mask = "A.FR.BCA_BP6_EUR");
# or
julia> df1 = rdb("IMF", "BOP", "A.FR.BCA_BP6_EUR");

julia> df2 = rdb("IMF", "BOP", mask = "A.FR+ES.BCA_BP6_EUR");

julia> df3 = rdb("IMF", "BOP", mask = "A..BCA_BP6_EUR");

julia> df4 = rdb("IMF", "BOP", mask = "A.FR.BCA_BP6_EUR+IA_BP6_EUR");


## By query
julia> df1 = rdb("IMF", "WEO:2019-10", query = "France current account balance percent");

julia> df2 = rdb("IMF", "WEO:2019-10", query = "current account balance percent");


## By api_link
julia> df1 = rdb(api_link = "https://api.db.nomics.world/v22/series/WB/DB?dimensions=%7B%22indicator%22%3A%5B%22IC.REG.PROC.FE.NO%22%5D%7D&q=Doing%20Business&observations=1&format=json&align_periods=1&offset=0&facets=0");
# or
julia> df2 = rdb("https://api.db.nomics.world/v22/series/WB/DB?dimensions=%7B%22indicator%22%3A%5B%22IC.REG.PROC.FE.NO%22%5D%7D&q=Doing%20Business&observations=1&format=json&align_periods=1&offset=0&facets=0");


## Use proxy with curl
julia> h = Dict(:proxy => "http://<proxy>:<port>");

julia> DBnomics.options("curl_config", h);
julia> df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN");
# or
julia> df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN", curl_config = h);

# Regarding the functioning of HTTP.jl, you might need to modify another option
# It will change the url from https:// to http://
# (https://github.com/JuliaWeb/HTTP.jl/pull/390)
julia> DBnomics.options("secure", false);


## Use readlines and download
julia> DBnomics.options("use_readlines", true);
julia> df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN");
# or
julia> df1 = rdb(ids = "AMECO/ZUTN/EA19.1.0.0.0.ZUTN", use_readlines = true);


## Apply a filter to the series
# One filter
julia> filters = Dict(:code => "interpolate", :parameters => Dict(:frequency => "daily", :method => "spline"));
julia> df1 = rdb(ids = ["IMF/WEO:2019-10/ABW.BCA.us_dollars", "IMF/WEO:2019-10/ABW.NGDPD.us_dollars"], filters = filters);

# For two filters
julia> filter1 = Dict(:code => "interpolate", :parameters => Dict(:frequency => "quarterly", :method => "spline"));
julia> filter2 = Dict(:code => "aggregate", :parameters => Dict(:frequency => "annual", :method => "average"));
julia> filters = (filter1, filter2);
julia> df1 = rdb(ids = ["IMF/WEO:2019-10/ABW.BCA.us_dollars", "IMF/WEO:2019-10/ABW.NGDPD.us_dollars"], filters = filters);

julia> filter1 = Dict(:code => "interpolate", :parameters => Dict(:frequency => "monthly", :method => "linear"));
julia> filter2 = Dict(:code => "x13", :parameters => nothing);
julia> filters = (filter1, filter2);
julia> df1 = rdb("ECB/EXR/A.AUD.EUR.SP00.A", filters = filters);
```
"""
function rdb(
    provider_code::Union{Array, String, Symbol, Nothing} = nothing,
    dataset_code::Union{String, Symbol, Nothing} = nothing,
    mask_arg::Union{String, Nothing} = nothing;
    ids::Union{Array, String, Nothing} = nothing,
    dimensions::Union{Dict, NamedTuple, String, Nothing} = nothing,
    mask::Union{String, Nothing} = nothing,
    query::Union{String, Nothing} = nothing,
    api_link::Union{String, Nothing} = nothing,
    use_readlines::Bool = DBnomics.use_readlines,
    curl_config::Union{Nothing, Dict, NamedTuple} = DBnomics.curl_config,
    filters::Union{Nothing, Dict, Tuple} = DBnomics.filters,
    kwargs...
)
    # Setting API url
    api_base_url::String = DBnomics.api_base_url
    # Setting API version
    api_version::Int64 = DBnomics.api_version
    # Setting API metadata
    metadata::Bool = DBnomics.metadata
    metadata_str::String = metadata ? "?" : ("?metadata=" * string(Int64(metadata)) * "&")
    
    api_base_url = api_base_url * "/v" * string(api_version) * "/series"
    
    # Checking arguments
    provider_code_null::Bool = isa(provider_code, Nothing)
    provider_code_not_null::Bool = !provider_code_null
    
    dataset_code_null::Bool = isa(dataset_code, Nothing)
    dataset_code_not_null::Bool = !dataset_code_null
    
    dimensions_null::Bool = isa(dimensions, Nothing)
    dimensions_not_null::Bool = !dimensions_null
    
    mask_null::Bool = isa(mask, Nothing)
    mask_not_null::Bool = !mask_null
    
    ids_null::Bool = isa(ids, Nothing)
    ids_not_null::Bool = !ids_null
    
    mask_arg_null::Bool = isa(mask_arg, Nothing)
    mask_arg_not_null::Bool = !mask_arg_null
    
    query_null::Bool = isa(query, Nothing)
    query_not_null::Bool = !query_null
    
    api_link_null::Bool = isa(api_link, Nothing)
    api_link_not_null::Bool = !api_link_null
    
    provider_code = isa(provider_code, Symbol) ? string(provider_code) : provider_code
    dataset_code = isa(dataset_code, Symbol) ? string(dataset_code) : dataset_code

    # provider_code is actually api_link
    if (
        provider_code_not_null && dataset_code_null &&
        dimensions_null && mask_null && ids_null && mask_arg_null &&
        query_null && api_link_null && DBnomics.rdb_no_arg
    )
        if isa(provider_code, String)
            if occursin(r"^http(s)*://", lowercase(provider_code))
                api_link = provider_code
                provider_code = nothing
                
                provider_code_null = true
                provider_code_not_null = !provider_code_null
                
                api_link_null = false
                api_link_not_null = !api_link_null
            end
        end
    end
    
    # By api_link
    if api_link_not_null
        return DBnomics.dot_rdb(
            api_link;
            use_readlines = use_readlines, curl_config = curl_config,
            filters = filters, kwargs...
        )
    end
    
    # provider_code is actually ids
    if (
        provider_code_not_null && dataset_code_null &&
        dimensions_null && mask_null && ids_null && mask_arg_null &&
        query_null && api_link_null && DBnomics.rdb_no_arg
    )
        ids = provider_code
        provider_code = nothing
        
        provider_code_null = true
        provider_code_not_null = !provider_code_null
        
        ids_null = false
        ids_not_null = !ids_null
    end
    
    # mask_arg is actually mask
    if (
        provider_code_not_null && dataset_code_not_null &&
        mask_arg_not_null && dimensions_null && mask_null &&
        ids_null && query_null && api_link_null && DBnomics.rdb_no_arg
    )
        mask = mask_arg
        mask_arg = nothing
        
        mask_null = false
        mask_not_null = !mask_null
        
        mask_arg_null = true
        mask_arg_not_null = !mask_arg_null
    end
    
    # By dimensions
    if dimensions_not_null
        if provider_code_null || dataset_code_null
            error(
                "When you filter with 'dimensions', you must specifiy " *
                "'provider_code' and 'dataset_code' as arguments of the " *
                "function."
            )
        end
        
        dimensions = to_json_if_dict_namedtuple(dimensions)
        
        link = api_base_url * "/" * provider_code * "/" * dataset_code *
            metadata_str * "&observations=1&dimensions=" * dimensions
        
        return DBnomics.dot_rdb(
            link;
            use_readlines = use_readlines, curl_config = curl_config,
            filters = filters, kwargs...
        )
    end
    
    # By mask
    if mask_not_null
        if provider_code_null || dataset_code_null
            error(
                "When you filter with 'mask', you must specifiy " *
                "'provider_code' and 'dataset_code' as arguments of the " *
                "function."
            )
        end
        
        link = api_base_url * "/" * provider_code * "/" * dataset_code *
            "/" * mask * metadata_str * "&observations=1"
        
        return DBnomics.dot_rdb(
            link;
            use_readlines = use_readlines, curl_config = curl_config,
            filters = filters, kwargs...
        )
    end
    
    # By ids
    if ids_not_null
        if provider_code_not_null || dataset_code_not_null
            if DBnomics.verbose_warning
                @warn "When you filter with 'ids', " *
                    "'provider_code' and 'dataset_code' are not considered."
            end
        end
        
        if isa(ids, String)
            ids = [ids]
        end
        if !isa(ids, Array{String, 1})
            error("'ids' must be an array of strings.")
        end
        if length(ids) <= 0
            error("'ids' is empty.")
        end
        
        link = api_base_url * metadata_str * "&observations=1&series_ids=" *
            reduce((u, w) -> u * "," * w, ids)
        
        return DBnomics.dot_rdb(
            link;
            use_readlines = use_readlines, curl_config = curl_config,
            filters = filters, kwargs...
        )
    end
    
    # By query
    if query_not_null
        if provider_code_null || dataset_code_null
            error(
                "When you filter with a 'query', you must specifiy " *
                "'provider_code' and 'dataset_code' as arguments of the " *
                "function."
            )
        end
        
        if DBnomics.verbose_warning
            if query == ""
                @warn "Your 'query' is empty, the entire dataset " *
                    provider_code * "/" * dataset_code *
                    " will be downloaded. It can be long !"
            end
        end
        
        link = api_base_url * "/" * provider_code * "/" * dataset_code *
          metadata_str * "&q=" * HTTP.escapeuri(query) * "&observations=1"
        
        return DBnomics.dot_rdb(
            link;
            use_readlines = use_readlines, curl_config = curl_config,
            filters = filters, kwargs...
        )
    end
    
    error("Please provide correct 'dimensions', 'mask', 'ids' or 'query'.")
end
